test/python/pulse/test_builder.py

```diff
@@ -0,0 +1,969 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2020.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Test pulse builder context utilities."""
+
+from math import pi
+import unittest
+
+import numpy as np
+
+from qiskit import circuit
+from qiskit import compiler
+from qiskit import pulse
+from qiskit.pulse import builder
+from qiskit.pulse import exceptions
+from qiskit.pulse import macros
+from qiskit.pulse import transforms
+from qiskit.pulse.instructions import directives
+from qiskit.test import QiskitTestCase
+from qiskit.test.mock import FakeOpenPulse2Q
+from qiskit.pulse import pulse_lib, instructions
+
+# pylint: disable=invalid-name
+
+
+class TestBuilder(QiskitTestCase):
+    """Test the pulse builder context."""
+
+    def setUp(self):
+        self.backend = FakeOpenPulse2Q()
+        self.configuration = self.backend.configuration()
+        self.defaults = self.backend.defaults()
+        self.inst_map = self.defaults.instruction_schedule_map
+
+
+class TestContexts(TestBuilder):
+    """Test builder contexts."""
+    def test_align_sequential(self):
+        """Test the sequential alignment context."""
+        d0 = pulse.DriveChannel(0)
+        d1 = pulse.DriveChannel(1)
+
+        with pulse.build() as schedule:
+            with pulse.align_sequential():
+                pulse.delay(3, d0)
+                pulse.delay(5, d1)
+                pulse.delay(7, d0)
+
+        reference = pulse.Schedule()
+        # d0
+        reference.insert(0, instructions.Delay(3, d0), mutate=True)
+        reference.insert(8, instructions.Delay(7, d0), mutate=True)
+        # d1
+        reference.insert(3, instructions.Delay(5, d1), mutate=True)
+
+        self.assertEqual(schedule, reference)
+
+    def test_align_left(self):
+        """Test the left alignment context."""
+        d0 = pulse.DriveChannel(0)
+        d1 = pulse.DriveChannel(1)
+        d2 = pulse.DriveChannel(2)
+
+        with pulse.build() as schedule:
+            with pulse.align_left():
+                pulse.delay(11, d2)
+                pulse.delay(3, d0)
+                with pulse.align_left():
+                    pulse.delay(5, d1)
+                    pulse.delay(7, d0)
+
+        reference = pulse.Schedule()
+        # d0
+        reference += instructions.Delay(3, d0)
+        reference += instructions.Delay(7, d0)
+        # d1
+        reference = reference.insert(3, instructions.Delay(5, d1))
+        # d2
+        reference += instructions.Delay(11, d2)
```

Comment:
preference for `insert` to make start times explicit everywhere

Reply:
Please see other comment for why I disagree.

Reply:
Fixed in all places where the timing of all reference schedules is nontrivial in 2a0130f04.